apps/serdes/Deserializer.cpp

```diff
@@ -0,0 +1,608 @@
+        auto args_serialized = call_expr->args();
+        std::vector<Halide::Expr> args;
+        for (size_t i = 0; i < args_serialized->size(); ++i) {
+            auto arg = deserialize_expr(args_type->Get(i), args_serialized->Get(i));
+            args.push_back(arg);
+        }
+        auto value_index = call_expr->value_index();
+        // TODO: fix type and function ptr here once function DAG is fixed
+        return Halide::Internal::Call::make(Halide::Int(64), name, args, Halide::Internal::Call::CallType::Extern, Halide::Internal::FunctionPtr(), value_index);
+    }
+    case Halide::Serdes::Expr::Expr_Variable: {
+        const Halide::Serdes::Variable *variable_expr = (const Halide::Serdes::Variable *)expr;
+        auto name = deserialize_string(variable_expr->name());
+        return Halide::Internal::Variable::make(Halide::Int(64), name);
+    }
+    case Halide::Serdes::Expr::Expr_Shuffle: {
+        const Halide::Serdes::Shuffle *shuffle_expr = (const Halide::Serdes::Shuffle *)expr;
+        auto vectors_type = shuffle_expr->vectors_type();
+        auto vectors_serialized = shuffle_expr->vectors();
+        std::vector<Halide::Expr> vectors;
+        for (size_t i = 0; i < vectors_serialized->size(); ++i) {
+            auto vector = deserialize_expr(vectors_type->Get(i), vectors_serialized->Get(i));
+            vectors.push_back(vector);
+        }
+        auto indices_serialized = shuffle_expr->indices();
+        std::vector<int32_t> indices;
+        for (size_t i = 0; i < indices_serialized->size(); ++i) {
+            indices.push_back(indices_serialized->Get(i));
+        }
+        return Halide::Internal::Shuffle::make(vectors, indices);
+    }
+    case Halide::Serdes::Expr::Expr_VectorReduce: {
+        const Halide::Serdes::VectorReduce *vector_reduce_expr = (const Halide::Serdes::VectorReduce *)expr;
+        auto value = deserialize_expr(vector_reduce_expr->value_type(), vector_reduce_expr->value());
+        // TODO: fix op here and store lanes during serialization
+        return Halide::Internal::VectorReduce::make(Halide::Internal::VectorReduce::Operator::Add, value, 16);
+    }
+    default: {
+        std::cerr << "unknown type code " << type_code << "\n";
+        return Halide::Expr();
+    }
+    }
+}
+
+Halide::Range Deserializer::deserialize_range(const Halide::Serdes::Range *range) {
+    auto min = deserialize_expr(range->min_type(), range->min());
+    auto extent = deserialize_expr(range->extent_type(), range->extent());
+    return Halide::Range(min, extent);
+}
+
+Pipeline Deserializer::deserialize(const std::string &filename) {
+    // unpack binary file
+    std::ifstream in(filename, std::ios::binary | std::ios::in);
+    if (!in) {
+        std::cerr << "failed to open file " << filename << "\n";
+        return Pipeline();
+    }
+    in.seekg(0, std::ios::end);
+    int size = in.tellg();
+    in.seekg(0, std::ios::beg);
+    char *data = new char[size];
```

Comment:
This is never freed. Does `GetPipeline()` take ownership below? If not, this is a leak.

If GetPipeline *does* take ownership, that needs a comment.

If GetPipeline *doesn't* take ownership, it would be cleaner to do something like

```
std::vector data(size);
in.read(data.data(), size);
```

Reply:
Thanks for finding this!